### ISSUE TITLE
rising_factorial and falling_factorial are defined weirdly

### ISSUE THREAD

**Issue by bgoodri** [stan-dev/math#620]
#### Summary:

The definitions of `(log_)rising_factorial` and `(log_)falling_factorial` in Stan are inconsistent with their mathematical definitions.

#### Description:

Stan restricts `x` to be positive in `rising_factorial` and `falling_factorial` but this restriction is not required.

Also, Stan allows `n` to be a real number but `rising_factorial` and `falling_factorial` are defined for non-negative integer values of `n`.

Also, `log_rising_factorial` and `log_falling_factorial` accept non-integer values of `n`, although in that case the restriction that `x` be positive is reasonable.

#### Reproducible Steps:

RTFM

#### Current Output:

```
Error in rising_factorial(-pi, 1) : 
  Exception: log_rising_factorial: first argument is -3.14159, but must be > 0!```
```
rising_factorial(2, pi) # 29.77011
```
#### Expected Output:

-pi and domain error respectively

#### Additional Information:


#### Current Version:
v2.17.0

**Comment by andrjohns** [comment]
It looks like the restrictions in rising/falling_factorial stem from them returning the exponentiated corresponding log_*_factorial, which doesn't exist when the factorial is negative. 

A possible solution could be to use the log_*_factorial when ```x``` is positive (to better handle under/overflow), and to use Boost's *_factorial when ``x``` is negative:
```
    template<typename T1, typename T2>
    inline typename boost::math::tools::promote_args<T1, T2>::type
    rising_factorial(const T1 x, const T2 n) {
      using std::exp;

      if (x < 0)
        return boost::math::rising_factorial(x, n, boost_policy_t());
    
      return exp(log_rising_factorial(x, n));
    }
```

**Comment by bgoodri** [comment]
I would say to use Boost's implementation unconditionally.

**Comment by andrjohns** [comment]
Easy enough. And both rising_ and falling_factorial should throw if ```n``` is a non-negative integer?

**Comment by bgoodri** [comment]
I think you meant throw if n is not a non-negative integer. In other words,
it should throw with a positive real number. This would require some
backwards incompatible changes to the Stan language though.

**Comment by andrjohns** [comment]
Wouldn't a conditional do the trick? Like:
```
    template<typename T1, typename T2>
    inline typename boost::math::tools::promote_args<T1, T2>::type
    rising_factorial(const T1 x, const T2 n) {

      if (!is_nonpositive_integer(x)){
        std::stringstream msg;
        msg << "x is " << x << " but x must be a non-negative integer!" << std::endl;
        throw std::domain_error(msg.str());
      }
     return boost::math::rising_factorial(x, n, boost_policy_t());
    }
```

**Comment by bgoodri** [comment]
I think it would be better to fail before runtime by making it be

T
rising_factorial(const T x, const int n) {
  check_nonnegative(n);
  boost::math::rising_factorial(x, n, boost_policy_t());
}

and then change the Stan language definitions so that it does not parse if
someone tries to pass a non-int for n.

**Comment by andrjohns** [comment]
I see what you mean. Should I introduce those as pull requests to the stan and math repos, or leave it since it would break backwards compatibility?

**Comment by bgoodri** [comment]
I don't think backward compatibility is a priority when the only things
that would get seriously broken are mathematically wrong. But someone else
should opine on this before we do it.

**Comment by syclik** [comment]
I'm with @bgoodri. I see this more as a bug rather than an intentional
behavior that that we had intended to maintain.

Now... there might (possibly) be places that call these functions that
should be doing additional checks.

**Comment by bob-carpenter** [comment]
I also agree.  This doesn't break backward compatibility---it fixes a bug.

**Comment by bgoodri** [comment]
OK. @andrjohns if you want to take this, go ahead and change the functions in Stan Math to call Boost's versions unconditionally, and then we will open up an issue to change the Stan language once that gets merged.
